python/tk_multi_launchapp/launch_apps.py

```diff
@@ -0,0 +1,651 @@
+            )
+
+    def launch_from_path(self, path, version=None):
+        """
+        Entry point if you want to launch an app given a particular path.
+        Note that there are no checks that the path passed is actually compatible
+        with the app that is being launched. This should be handled in logic
+        which is external to this app.
+        """
+        context = self._tk_app.sgtk.context_from_path(path)
+        self._launch_app(context, path, version)
+
+    def launch_from_path_and_context(self, path, context, version=None):
+        """
+        Extended version of launch_from_path. This method takes an additional
+        context parameter which is useful if you want to seed the launch context
+        with more context data than is available in the path itself. Typically
+        paths may not contain a task, so this may need to be pushed through
+        separately via the context.
+        """
+        if context is None:
+            # this context looks sour. So fall back on to path-only launch.
+            self.launch_from_path(path, version)
+        else:
+            # use given context to launch engine!
+            self._launch_app(context, path, version)
+
+
+class SoftwareEntityLauncher(BaseLauncher):
+    """
+    Launches a DCC based on site Software entity entries.
+    """
+    def init_launch_commands(self):
+        """
+        Determine what launch command(s) to register with the current TK engine.
+        Multiple commands may be registered based on the number of retrieved
+        Software entities and their corresponding 'versions' field.
+        """
+        # Expand Software field names that rely on the current platform
+        app_path_field = "sg_%s_path" % self._platform_name
+        app_args_field = "sg_%s_args" % self._platform_name
+
+        # Determine the information to retrieve from SG
+        sw_entity = self._tk_app.get_setting("sg_software_entity") or "Software"
+        sw_filters = [
+            ["sg_status_list", "is", "act"],
+        ]
+        sw_fields = [
+            app_path_field,
+            app_args_field,
+            "code",
+            "image",
+            "sg_engine",
+            "sg_group_restrictions",
+            "sg_projects",
+            "sg_versions",
+            "sg_user_restrictions",
+        ]
+
+        # Get the list of Software apps that match the specified filters.
+        sw_entities = self._tk_app.shotgun.find(
+            sw_entity, sw_filters, sw_fields
+        )
+        if not sw_entities:
+            # No Entities found matching filters, nothing to do.
+            self._tk_app.log_debug("No SG %s entities found matching filters : %s" %
+                (sw_entity, sw_filters)
+            )
+            return
+
+        # Short-hand for a couple of useful context values
+        ctx_user = self._tk_app.context.user
+        ctx_project = self._tk_app.context.project
+
+        for sw_app in sw_entities:
```

Comment:
for debug and support purposes (i am thinking esp. trying to figure out group/user/project constraints not working as they should), i think it could be useful to print some debug here to show what was received from Shotgun